x-pack/plugins/security_solution/public/management/services/exceptions_list/exceptions_list_api_client.ts

```diff
@@ -0,0 +1,193 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import {
+  CreateExceptionListItemSchema,
+  CreateExceptionListSchema,
+  ExceptionListItemSchema,
+  ExceptionListSummarySchema,
+  FoundExceptionListItemSchema,
+  UpdateExceptionListItemSchema,
+} from '@kbn/securitysolution-io-ts-list-types';
+import { EXCEPTION_LIST_ITEM_URL, EXCEPTION_LIST_URL } from '@kbn/securitysolution-list-constants';
+import { HttpStart } from 'kibana/public';
+
+/**
+ * A generic class to be used for each artifact type.
+ * It follow the Singleton pattern.
+ * Please, use the getInstance method instead of creating a new instance when using this implementation.
+ */
+export class ExceptionsListApiClient {
+  private static instance: ExceptionsListApiClient;
+  private ensureListExists: Promise<void>;
+
+  constructor(
+    private readonly http: HttpStart,
+    private readonly listId: string,
+    private readonly listDefinition: CreateExceptionListSchema
+  ) {
+    this.ensureListExists = this.createExceptionList();
+  }
+
+  /**
+   * PrivateStatic method that creates the list and don't throw if list already exists.
+   * This method is being used when initializing an instance only once.
+   */
+  private async createExceptionList(): Promise<void> {
+    try {
+      await this.http.post<ExceptionListItemSchema>(EXCEPTION_LIST_URL, {
+        body: JSON.stringify({ ...this.listDefinition, list_id: this.listId }),
+      });
+    } catch (err) {
+      // Ignore 409 errors. List already created
+      if (err.response?.status !== 409) {
+        throw err;
+      }
+    }
+  }
+
+  /**
+   * Static method to get a fresh or existing instance.
+   * It will ensure we only check and create the list once.
+   */
+  public static getInstance(
+    http: HttpStart,
+    listId: string,
+    listDefinition: CreateExceptionListSchema
+  ): ExceptionsListApiClient {
+    if (!ExceptionsListApiClient.instance) {
```

Comment:
I don't think this is a good idea here because the first time someone uses the `getInstance()` method, that instantiation will stick. Example: 

```typescript
// call 1: 
const trustedAppsInstance = ExceptionsListApiClient.getInstance(
  coreHttp, 
  'trusted_apps', 
  trustedAppListDefinition
);


// call 2:
const eventFilters = ExceptionsListApiClient.getInstance(
  coreHttp, 
  'event_filters', 
  eventFiltersDefinition
);
// the above would actually return the Trusted Apps instance since that was the first one to be created
```

Reply:
oh, that's totally right. I forgot to set this by list id. What do you think about having the instance as a Map of key values where the key is the list_id and the value is the Instance for that list id?
Something like:

```
  private static instance: Map<ListId, ExceptionsListApiClient>;

  public static getInstance(
    http: HttpStart,
    listId: string,
    listDefinition: CreateExceptionListSchema
  ): ExceptionsListApiClient {
    if (!ExceptionsListApiClient.instance.has(listId)) {
      ExceptionsListApiClient.instance.set(
        listId,
        new ExceptionsListApiClient(http, listId, listDefinition)
      );
    }
    const currentInstance = ExceptionsListApiClient.instance.get(listId);
    if (currentInstance) {
      return currentInstance;
    } else {
      return new ExceptionsListApiClient(http, listId, listDefinition);
    }
  }
```

This will ensure each listId uses it's own class instance

Reply:
This will also work. So your idea here is that `.getInstance()` will create a class instance for a given list type and store that instance to be reused, right? I think that ok and it matches an idea that had been floating in the back of my mind to create a React context for Artifact APIs. this is even better because we can use it outside of React if needed.

I don't know why, but it feels strange to me to store mutating data on the Class its self - even when its a private static attribute

Reply:
Yes, that's exactly the purpose of this method. If I want to get a fresh instance, I can just create the constructor, but if I want to get an instance that was previously created, then I can use the getInstance method that will give me an existing instance if exists or a new one.
And yes! We can use this outside of react context.

I guess this is the way to follow the Singleton patter in TS class context, but let me know if you think anything could be done different. I will push the changes so you can check all together

Reply:
This is good. 👍